driver/main.c

```diff
@@ -157,169 +219,447 @@ PetscErrorCode SetConstantRainfall(PetscReal rain_rate, PetscInt ncells, PetscRe
+    PetscReal yc = data->mesh_yc[icell];
+
+    PetscReal min_dist;
+    for (PetscInt kk = 0; kk < data->ndata; kk++) {
+      PetscReal dx   = xc - data->data_xc[kk];
+      PetscReal dy   = yc - data->data_yc[kk];
+      PetscReal dist = PetscPowReal(dx * dx + dy * dy, 0.5);
+
+      if (kk == 0) {
+        min_dist                   = dist;
+        data->data2mesh_idx[icell] = kk;
+      } else {
+        if (dist < min_dist) {
+          min_dist                   = dist;
+          data->data2mesh_idx[icell] = kk;
+        }
+      }
+    }
+
+    if (0) {
+      PetscInt idx = data->data2mesh_idx[icell];
+      printf("%04" PetscInt_FMT " %f %f %02" PetscInt_FMT " %f %f\n", icell, xc, yc, idx, data->data_xc[idx], data->data_yc[idx]);
+    }
+  }
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+/// @brief Closes the currently open Vec and loads in the next binary file in the Vec
+/// @param data A pointer to UnstructuredDataset
+/// @return PETSC_SUCESS on success
+static PetscErrorCode OpenANewUnstructuredDataset(UnstructuredDataset *data) {
+  PetscFunctionBegin;
+
+  // close the existing file
+  PetscCall(VecRestoreArray(data->data_vec, &data->data_ptr));
+  PetscCall(VecDestroy(&data->data_vec));
+
+  // increase the date
+  struct tm *current_date = &data->current_date;
+  current_date->tm_hour++;
+  mktime(current_date);
+
+  // determine the new file
+  PetscCall(DetermineDatasetFilename(&data->current_date, data->dir, data->file));
+  PetscPrintf(PETSC_COMM_WORLD, "Opening %s \n", data->file);
+
+  PetscViewer viewer;
+  PetscCall(VecCreate(PETSC_COMM_SELF, &data->data_vec));
+  PetscCall(PetscViewerBinaryOpen(PETSC_COMM_SELF, data->file, FILE_MODE_READ, &viewer));
+  PetscCall(VecLoad(data->data_vec, viewer));
+  PetscCall(PetscViewerDestroy(&viewer));
+
+  PetscCall(VecGetArray(data->data_vec, &data->data_ptr));
+
+  data->ndata_file++;
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+/// @brief Closes the currently open Vec and free up memory used for mapping
+/// @param data A pointer to UnstructuredDataset struct
+/// @return PETSC_SUCESS on success
+static PetscErrorCode DestroyUnstruturedcDataset(UnstructuredDataset *data) {
+  PetscFunctionBegin;
+
+  // close the existing file
+  PetscCall(VecRestoreArray(data->data_vec, &data->data_ptr));
+  PetscCall(VecDestroy(&data->data_vec));
+
+  PetscCall(PetscFree(data->data_xc));
+  PetscCall(PetscFree(data->data_yc));
+  PetscCall(PetscFree(data->mesh_xc));
+  PetscCall(PetscFree(data->mesh_yc));
+  PetscCall(PetscFree(data->data2mesh_idx));
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+/// @brief Set spatially varying BC conditions from an unstructured dataset
+/// @param data        [in]  Pointer to a UnstructuredDataset struct
+/// @param cur_time    [in]  Current time
+/// @param data_values [out] BC values for boundary cells
+/// @return PETSC_SUCESS on success
+PetscErrorCode SetUnstructuredDataset(UnstructuredDataset *data, PetscReal cur_time, PetscReal *data_values) {
```

Comment:
Maybe we can call this `SetUnstructuredData` to emphasize that the data is being taken _from_ the dataset and placed into `data_values`.